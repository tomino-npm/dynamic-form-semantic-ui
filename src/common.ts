import { FormElement, DataSet } from '@tomino/dynamic-form';
import { initCss } from '@tomino/toolbelt';

export type SignatureType = {
  comment: string;
  signature: string;
  verifiedState: 'Pending' | 'Verified' | 'Rejected';
  rejected?: boolean;
  uid: string;
  name: string;
  date: Date;
  setValue?(name: string, value: any): any;
};

export type SignatureHandlers = {
  validateForm(): boolean;
  verifySignature(uid: string, signature: string): Promise<string>;
  sign(
    source: string,
    reject: boolean,
    password: string,
    reason: string,
    submit: boolean
  ): Promise<SignatureType>;
  signatureFont(): string;
};

export type FormControlProps = {
  formControl: FormElement;
  owner: DataSet;
  handlers?: { [index: string]: any };
  readOnly: boolean;
};

export const { css, renderCss } = initCss('form-');
