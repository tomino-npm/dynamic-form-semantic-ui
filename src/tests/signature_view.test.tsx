import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { create } from './form_query_data';
import { JSONSchema, FormDefinition, FormModel, config } from '@tomino/dynamic-form';
import { TestComponent } from './common';

const date = new Date('2018-10-15T00:05:32.000Z');
const signatureSchema: JSONSchema = {
  type: 'object',
  properties: {
    comment: {},
    signature: {},
    verifiedState: {},
    rejected: {
      type: 'boolean'
    },
    uid: {},
    name: {},
    date: {
      type: 'string',
      format: 'date-time'
    }
  }
};
const schema: JSONSchema = {
  type: 'object',
  properties: {
    deanSignature: signatureSchema,
    supervisorSignature: signatureSchema,
    supervisorSignature2: signatureSchema,
    supervisorSignature3: signatureSchema,
    studentSignature: signatureSchema
  }
};

const signature =
  'oiuoiu3o4uoi34u43o34kbkjh43jhbiu3y498734h34ln34oiu349873948u3ol4n3l4n3oi4j3o4u394879384yo3hj4oij34o3u94873984739843iou4hk3j4hio3uh4iu3y497834987349873984ho34ihj';
const controlData = {
  deanSignature: {
    comment: 'All wrong',
    date,
    rejected: true,
    name: 'Simeon Simoff'
  },
  supervisorSignature: {
    date,
    signature,
    name: 'Alladin Hasan'
  },
  supervisorSignature2: {
    date,
    signature,
    name: 'Alladin Hasan'
  },
  supervisorSignature3: {
    date,
    signature,
    name: 'Alladin Hasan'
  },
  studentSignature: {}
};

const formDefinition: FormDefinition = create.form({
  elements: [
    create.formElement({
      row: 0,
      column: 0,
      width: 5,
      control: 'Signature',
      label: 'Dean Signature',
      source: 'deanSignature',
      controlProps: {
        font: 'Pacifico',
        allowComment: false
      }
    }),
    create.formElement({
      row: 0,
      column: 5,
      width: 5,
      control: 'Signature',
      label: 'Supervisor Signature',
      source: 'supervisorSignature',
      controlProps: {
        font: 'Indie Flower',
        allowComment: false
      }
    }),
    create.formElement({
      row: 0,
      column: 10,
      width: 5,
      control: 'Signature',
      label: 'Student Signature',
      source: 'studentSignature',
      controlProps: {
        font: 'Charm',
        allowComment: false
      }
    }),
    create.formElement({
      row: 1,
      column: 10,
      width: 5,
      control: 'Signature',
      label: 'Student Signature',
      source: 'studentSignature',
      controlProps: {
        font: 'Charm',
        allowComment: true,
        submit: true
      }
    }),
    create.formElement({
      row: 1,
      column: 5,
      width: 5,
      control: 'Signature',
      label: 'Supervisor Signature',
      source: 'supervisorSignature2',
      controlProps: {
        font: 'Indie Flower',
        allowComment: false
      }
    }),
    create.formElement({
      row: 2,
      column: 5,
      width: 5,
      control: 'Signature',
      label: 'Supervisor Signature',
      source: 'supervisorSignature3',
      controlProps: {
        font: 'Indie Flower',
        allowComment: false
      }
    }),
    create.formElement({
      row: 2,
      column: 10,
      width: 5,
      control: 'Signature',
      label: 'Student Signature',
      source: 'studentSignature',
      controlProps: {
        font: 'Charm',
        allowComment: false,
        allowReject: true
      }
    }),
    create.formElement({
      row: 3,
      column: 10,
      width: 5,
      control: 'Signature',
      label: 'Student Signature',
      source: 'studentSignature',
      controlProps: {
        font: 'Charm',
        allowComment: true,
        allowReject: true
      }
    })
  ]
});

const results = ['Pending', 'Verified', 'Rejected'];
let ci = 0;

describe('Signature', function() {
  const setDirty = jest.fn();

  beforeAll(() => {
    config.setDirty = () => setDirty();
  });

  function componentWithData() {
    const form = new FormModel(formDefinition, schema, controlData);
    const handlers = {
      sign: () => {
        return new Promise(resolve =>
          setTimeout(() => {
            resolve({
              date,
              signature,
              name: 'Alladin Hasan',
              verifiedState: 'Verified'
            } as any);
          }, 1000)
        );
      },
      verifySignature: () => Promise.resolve(results[ci++ % 3]),
      signatureFont: () => 'Pacifico',
      validateForm: () => true
    };

    // just another notation
    return <TestComponent form={form} handlers={handlers} />;
  }

  it('renders correctly', () => {
    const component = renderer.create(componentWithData());
    expect(component).toMatchSnapshot();
  });

  // it('changes value and all related formulas', () => {
  //   const component = renderer.create(componentWithData());
  //   const root = component.root;
  //   const country = root.findByProps({ name: 'country' });
  //   country.props.onChange(null, { value: 'AU' });

  //   const city = root.findByProps({ name: 'city' });
  //   city.props.onChange(null, { value: 'SYD' });

  //   expect(component).toMatchSnapshot();
  //   expect(setDirty).toBeCalled();
  // });

  return { componentWithData };
});
