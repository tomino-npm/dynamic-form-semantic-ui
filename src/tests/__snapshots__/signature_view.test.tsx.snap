// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Signature renders correctly 1`] = `
<div
  className="ui segment ui form"
>
  <style>
    
  </style>
  <div
    className="ui form"
  >
    <div
      className="fields"
    >
      <div
        className="five wide field"
      >
        <fieldset
          className="signature-Abfgejbdfga"
        >
          <legend>
            Dean Signature
          </legend>
          <div
            className="ui list"
            role="list"
          >
            <div
              className="item"
              onClick={[Function]}
              role="listitem"
            >
              <i
                aria-hidden="true"
                className="ban icon"
              />
              <div
                className="content"
              >
                Simeon Simoff rejected to sign on 10/15/2018
              </div>
            </div>
            <div
              className="item"
              onClick={[Function]}
              role="listitem"
            >
              <i
                aria-hidden="true"
                className="comment icon"
              />
              <div
                className="content"
              >
                All wrong
              </div>
            </div>
          </div>
        </fieldset>
      </div>
      <div
        className="five wide field"
      >
        <fieldset
          className="signature-Abfgejbdfga"
        >
          <legend>
            Supervisor Signature
          </legend>
          
        </fieldset>
      </div>
      <div
        className="five wide field"
      >
        <fieldset
          className="signature-Abfgejbdfga"
        >
          <legend>
            Student Signature
          </legend>
          
          <button
            className="ui icon primary left labeled button"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <i
              aria-hidden="true"
              className="pencil alternate icon"
            />
            Sign
          </button>
        </fieldset>
      </div>
    </div>
    <div
      className="fields"
    >
      <div
        className="five wide field"
      >
         
      </div>
      <div
        className="five wide field"
      >
        <fieldset
          className="signature-Abfgejbdfga"
        >
          <legend>
            Supervisor Signature
          </legend>
          
        </fieldset>
      </div>
      <div
        className="five wide field"
      >
        <fieldset
          className="signature-Abfgejbdfga"
        >
          <legend>
            Student Signature
          </legend>
          
          <textarea
            onChange={[Function]}
            onInput={[Function]}
            placeholder="Please leave a comment ..."
            rows={3}
            style={
              Object {
                "resize": "",
              }
            }
            value=""
          />
          <button
            className="ui icon primary right floated left labeled button"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <i
              aria-hidden="true"
              className="pencil alternate icon"
            />
            Sign
          </button>
        </fieldset>
      </div>
    </div>
    <div
      className="fields"
    >
      <div
        className="five wide field"
      >
         
      </div>
      <div
        className="five wide field"
      >
        <fieldset
          className="signature-Abfgejbdfga"
        >
          <legend>
            Supervisor Signature
          </legend>
          
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
